appscripts/canalytics.gs

```diff
@@ -0,0 +1,312 @@
+ *   ( reference: https://www.thebluealliance.com/apidocs look for Caching )
+ *   If the response code is 304, then the data has not changed, and the Cache is still valid even though the time has passed.
+ *   Since the data can be valid past the max-age, we lie to the cache-control and ask it to save data for 6 hours.
+ * 
+ * @param {text} path The part of the URL after "/api/v3/" in the query.
+ * @return Parsed json of the result including a cache time and Etag used for reducing data transfer.
+ * @customfunction
+ */
+function TBAQuery(path) {
+  var url = 'https://www.thebluealliance.com/api/v3/'+path
+  const cacheDocument = CacheService.getDocumentCache();
+  var headers = {}
+  
+  if ( getAPIKey_() === null ){
+    throw new Error("No API Key, set key before running any other scripts.");
+  }else{
+    headers['X-TBA-Auth-Key'] = getAPIKey_()
+  }
+  
+  var cacheStats = JSON.parse(cacheDocument.get("cacheStats" + url ))
+  var cacheResult = cacheDocument.get(url)
+
+  if ( cacheStats != null && cacheResult != null ){
+    // Data is cached.  Let's check to see if it is still good
+    // Cache Service can decide to arbitrarily remove cache entries, so check both.
+    var now = new Date().getTime();
+    Logger.log("url: " + url + " cacheExpireMs: " + cacheStats.cacheExpireMs + " ETag: " + cacheStats.etag + "date: " + now )
+    if ( cacheStats.cacheExpireMs > now ){
+      Logger.log("Within max age, returning cache")
+      return cacheResult;
+    }else{
+       headers['If-None-Match'] = cacheStats.etag
+    }
+  }
+  // headers['If-None-Match'] = "900cc8acb24b1c43ba74d061345260c317dcd610" for https://www.thebluealliance.com/api/v3//event/2023txhou/teams/simple
+
+  var params = {
+    'headers' : headers,
+    'muteHttpExceptions' : true
+  }
+
+  var result = UrlFetchApp.fetch(url, params);
+  Logger.log("TBA response code:" + result.getResponseCode())
+  if ( result.getResponseCode() === 304 ){
+    // Even though the data is past the cache timeout, TBA has confirmed that nothing has changed.
+    // Return the stale cache.
+    Logger.log(result.getHeaders())
+    Logger.log("nothing changed returned from tba, returning stale cache")
+    Logger.log(cacheResult)
+    return cacheResult
+  }else if ( result.getResponseCode() != 200 ){
+    throw new Error("Failed TBA call.  URL: " + url + " Response code: " + result.getResponseCode() )
+  }
+  
+  // Good return status, cache locally cache stats and data.
+  const resultHeaders = result.getHeaders()
+  var maxAge = resultHeaders["Cache-Control"].match(/max-age=(\d+)/)[1]
+  
+  var jsonResult = JSON.parse(result.getContentText());
+  var cacheStats = {}
+
+  cacheStats['cacheExpireMs'] = new Date().getTime() + maxAge*1000,
+  cacheStats['etag'] = resultHeaders['ETag']
+  
+  cacheDocument.put("cacheStats" + url, JSON.stringify(cacheStats), 21600)
+  cacheDocument.put(url, jsonResult, 21600)
```

Comment:
Getting data from cache doesn't seem work correctly (i.e. using an event key a second time doesn't work). I think the jsonResult needs to be converted to a string like the cacheStats.

Reply:
Thanks, will look further into it.  storing a string doesn't work because it is over 100K.

Reply:
Found a good chunk function and used it to store multiple cache items.